Use (#587)
reactjs/fr.react.dev: src/content/reference/react/use.md

<Canary>

> Le Hook `use` n'est actuellement disponible que sur les canaux de distribution canari et expérimentaux de React. Apprenez-en davantage sur [les canaux de versions de React ici](/community/versioning-policy#all-release-channels).

Author: Il faudra penser à mettre à jour la traduction du HTML généré par le composant `<Canary>`, car il indique "Canari (fonctionnalité expérimentale)"

  if (show) {
    return <MessageContainer messagePromise={messagePromise} />;
  } else {
    return <button onClick={download}>Télécharger le message</button>;
  }
}
```

```js index.js hidden
// TODO : remplacer l'import d'une version canary
// de React par une version stable, dès qu'elle
// intégrera le Hook `use`
import React, { StrictMode } from 'react';
import { createRoot } from 'react-dom/client';
import './styles.css';

// TODO : mettre à jour cet exemple pour utiliser
// le composant serveur Codesandbox de l'environnement

> // de démo lorsqu'il sera créé.

Author: J'avoue avoir hésité à traduire ces `TODO`, ils donnent l'impression de s'adresser aux développeurs de React directement...

> // de démo lorsqu'il sera créé.

Reviewer: En fait il y a une syntaxe dans MDX qui te permet de pondre un commentaire qui n'est pas *renderé* sans que ce soit un commentaire HTML (interdit en MDX). Elle est utilisée dans `cache` notamment. Là s'il l'a foutu dans son exemple, OK, mais on traduit ! 😉

}
```

> Toutefois, l'utilisation du `await` dans un [composant serveur](/reference/react/components#server-components) bloquera le rendu jusqu'à ce que l'instruction `await` soit terminée. Le passage d'une promesse d'un composant serveur à un composant client permet à la promesse de ne pas bloquer le rendu du composant serveur.

Reviewer: ```suggestion
Toutefois, l'utilisation d'`await` dans un [Composant Serveur](/reference/react/components#server-components) bloquera son rendu jusqu'à ce que l'instruction `await` ait terminé. Le passage d'une promesse d'un Composant Serveur à un Composant Client permet à la promesse de ne pas bloquer le rendu du Composant Serveur.
```

</DeepDive>

### Traiter les promesses rejetées {/*dealing-with-rejected-promises*/}

Dans certains cas, une promesse passée à `use` peut être rejetée. Vous pouvez gérer les promesses rejetées en utilisant l'une ou l'autre de ces méthodes :

1. [afficher une erreur aux utilisateurs avec un périmètre d'erreur](#displaying-an-error-to-users-with-error-boundary)
2. [fournir une valeur alternative avec `Promise.catch`](#providing-an-alternative-value-with-promise-catch)

<Pitfall>

> `use` ne peut pas être appelé à l'intérieur d'un bloc try-catch. À la place de ce type de bloc, [enrobez votre composant dans un périmètre d'erreur](#displaying-an-error-to-users-with-error-boundary), ou [fournissez une valeur alternative à utiliser avec la méthode `.catch` des promesses](#providing-an-alternative-value-with-promise-catch).

Reviewer: ```suggestion
`use` ne peut pas être appelé à l'intérieur d'un bloc try-catch. En remplacement de ces blocs, [enrobez votre composant dans un périmètre d'erreur](#displaying-an-error-to-users-with-error-boundary), ou [fournissez une valeur alternative à utiliser avec la méthode `.catch` des promesses](#providing-an-alternative-value-with-promise-catch).
```

</Pitfall>

#### Afficher une erreur aux utilisateurs dans un périmètre d'erreur {/*displaying-an-error-to-users-with-error-boundary*/}

> Si vous souhaitez afficher une erreur à vos utilisateurs quand une promesse a été rejetée, vous pouvez utiliser un [périmètre d'erreur](/reference/react/Component#catching-rendering-errors-with-an-error-boundary). Pour l'utiliser, enrobez le composant d'où vous appelez le Hook `use` dans le périmètre d'erreur. Si la promesse transmise à `use` est rejetée, alors l'UI de secours de ce périmètre d'erreur est affichée.

Reviewer: ```suggestion
Si vous souhaitez afficher une erreur à vos utilisateurs quand une promesse a été rejetée, vous pouvez utiliser un [périmètre d'erreur](/reference/react/Component#catching-rendering-errors-with-an-error-boundary). Pour l'utiliser, enrobez le composant d'où vous appelez le Hook `use` dans le périmètre d'erreur. Si la promesse transmise à `use` est rejetée, alors l'UI de secours de ce périmètre d'erreur sera affichée.
```

export function MessageContainer({ messagePromise }) {
  return (

>     <ErrorBoundary fallback={<p>⚠️ Quelque chose s'est mal passé</p>}>

Reviewer: ```suggestion
    <ErrorBoundary fallback={<p>⚠️ Ça sent le pâté…</p>}>
```

Harmonisation de trad 😂

Author: Je trouve ça discriminatoire envers le pâté, mais bon...

Reviewer: Et pourtant j'adore le pâté !

  if (show) {
    return <MessageContainer messagePromise={messagePromise} />;
  } else {
    return <button onClick={download}>Télécharger le message</button>;
  }
}
```

```js index.js hidden

> // TODO : remplacer l'import d'une version canary
> // de React par une version stable, dès qu'elle
> // intégrera le Hook `use`

Reviewer: ```suggestion
// TODO : remplacer l’import d’une version Canary
// de React par une version stable, dès qu’elle
// intégrera le Hook `use`
```

import React, { StrictMode } from 'react';
import { createRoot } from 'react-dom/client';
import './styles.css';

> // TODO : mettre à jour cet exemple pour utiliser
> // le composant serveur Codesandbox de l'environnement
> // de démo lorsqu'il sera créé.

Reviewer: ```suggestion
// TODO : mettre à jour cet exemple pour utiliser
// le Composant Serveur Codesandbox de l’environnement
// de démo lorsqu’il sera créé.
```

  const messagePromise = new Promise((resolve, reject) => {
    reject();
  }).catch(() => {
    return "Aucun nouveau message.";
  });

  return (
    <Suspense fallback={<p>En attente de message...</p>}>
      <Message messagePromise={messagePromise} />
    </Suspense>
  );
}
```

> Pour utiliser la méthode <CodeStep step={1}>`catch`</CodeStep> de la promesse, appelez <CodeStep step={1}>`catch`</CodeStep> sur l'objet `Promise`. <CodeStep step={1}>`catch`</CodeStep> n'accepte qu'un seul paramètre : une fonction qui prend un message d'erreur comme argument. Ce qui est <CodeStep step={2}>renvoyé</CodeStep> par la fonction passée à <CodeStep step={1}>`catch`</CodeStep> sera utilisé comme valeur résolue de la promesse.

Reviewer: ```suggestion
Pour utiliser la méthode <CodeStep step={1}>`catch`</CodeStep> de la promesse, appelez <CodeStep step={1}>`catch`</CodeStep> sur l'objet `Promise`. <CodeStep step={1}>`catch`</CodeStep> n'accepte qu'un seul paramètre : une fonction qui prend une erreur comme argument. La <CodeStep step={2}>valeur de retour</CodeStep> de la fonction passée à <CodeStep step={1}>`catch`</CodeStep> sera utilisée comme valeur accomplie de la promesse.
```

Erreur dans la VO (y'en a aussi des belles dans la VO de `catch`).

Hélas pas toujours le temps de faire des tickets *upstream*, mais je m'attends clairement à des conflits de sync le lundi sur ces nouvelles pages, car ils vont corriger / reformuler plein de trucs.

_C'est la vie_

---

## Dépannage {/*troubleshooting*/}

### *"Suspense Exception: This is not a real error!"* {/*suspense-exception-error*/}

> Vous appelez `use` soit en dehors d'un composant React soit d'une fonction de Hook, ou encore vous appelez `use` dans un bloc try-catch. Pour résoudre ce dernier cas, enrobez votre composant dans un périmètre d'erreur ou appelez la fonction `Promise.catch` pour attraper l'erreur et résoudre la promesse avec une valeur différente. [Consultez ces exemples](#dealing-with-rejected-promises).

Reviewer: ```suggestion
*(« Exception Suspense : ce n'est pas une véritable erreur ! », NdT)*

Vous appelez `use` soit en-dehors d'un composant React ou d'une fonction de Hook, soit dans un bloc try-catch. Pour corriger ce dernier cas, enrobez votre composant dans un périmètre d'erreur ou appelez la fonction `Promise.catch` pour attraper l'erreur et résoudre la promesse avec une valeur différente. [Consultez ces exemples](#dealing-with-rejected-promises).
```

> Si vous appelez `use` en dehors d'un composant React ou d'une fonction de Hook, déplacez l'appel à `use` dans un composant React ou une fonction Hook.

Reviewer: ```suggestion
Si vous appelez `use` en-dehors d'un composant React ou d'une fonction de Hook, déplacez l'appel à `use` dans un composant React ou une fonction de Hook.
```

Magique le conseil 😂

```jsx
function MessageComponent({messagePromise}) {
  function download() {
    // ❌ la fonction appelant `use` n'est ni un composant ni un Hook
    const message = use(messagePromise);
    // ...
```

> Appelez plutôt `use` en dehors de toute fermeture lexicale des composants, lorsque la fonction qui appelle `use` est un composant ou un Hook.

Reviewer: ```suggestion
Appelez plutôt `use` hors de toute fermeture lexicale au sein du composant, lorsque la fonction qui appelle `use` est un composant ou un Hook.
```
